stanza/text/vocab.py

```diff
@@ -134,33 +110,53 @@ def update(self, words):
         """
         return [self.add(w) for w in words]
 
-    def words2indices(self, words):
-        """
-        Convert a list of words into a list of indices.
+    def word2index(self, w):
+        return self[w]
 
-        :param words: an iterable of words to map to indices.
+    def index2word(self, i):
+        return self._index2word[i]
+
+    def freeze(self):
+        return FrozenVocab(self)
+
+    def count(self, w):
+        """Get the count for a word.
 
-        :return: the corresponding indices for each word. If a word is not found in the vocabulary then the unknown index will be returned for it.
+        :param w: a string
         """
-        return [self[w] for w in words]
+        return self._counts[w]
 
-    def indices2words(self, indices):
+    def counts_copy(self):
+        """
+        :return: a copy of the counter containing the number of occurrences of each word.
         """
-        Converts a list of indices into a list of words.
+        return copy(self._counts)
 
-        :param indices: indices for which to retrieve words.
+    def _refresh_index2word_cache(self):
+        try:
+            self._index2word_cache
+        except AttributeError:
+            # create if it doesn't exist
+            self._index2word_cache = self._compute_index2word()
 
-        :return: a list of words corresponding to each index.
-        """
-        index2word = self._index2word_copy()
-        return [index2word[i] for i in indices]
+        # update if it is out of date
+        if len(self._index2word_cache) != len(self):
+            self._index2word_cache = self._compute_index2word()
+
+    def _compute_index2word(self):
+        return self.keys()  # works because self is an OrderedDict
 
     @property
-    def counts(self):
-        """
-        :return: a counter containing the number of occurrences of each word.
+    def _index2word(self):
```

Comment:
I would actually just merge all the if-else cache check logic into this function call, instead of having 3 functions